spec/services/order_invoice_generator_spec.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe OrderInvoiceGenerator do
+  let!(:order) { create(:completed_order_with_fees) }
+  let!(:invoice_data_generator){ InvoiceDataGenerator.new(order) }
+  let!(:latest_invoice){
+    create(:invoice,
+           order:,
+           data: invoice_data_generator.serialize_for_invoice)
+  }
+
+  let(:instance) { described_class.new(order) }
+  let(:comparator){ double("OrderInvoiceComparator") }
+
+  before do
+    allow(instance).to receive(:comparator).and_return(comparator)
+  end
+
+  describe "#generate_or_update_latest_invoice" do
+    let(:subject) { instance.generate_or_update_latest_invoice }
+    context "when can generate new invoice" do
+      before do
+        expect(comparator).to receive(:can_generate_new_invoice?).and_return(true)
+      end
+
+      it "should create a new invoice" do
+        expect(instance).to receive(:invoice_data)
+        expect{ subject }.to change{ order.invoices.count }.by(1)
```

Comment:
Perhaps we should check that the invoice number was generated correctly too, but it's ok.